bundles/org.openhab.binding.clementineremote/src/main/java/org/openhab/binding/clementineremote/internal/ClementineRemoteHandler.java

```diff
@@ -0,0 +1,425 @@
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.qspool.clementineremote.backend.pb.ClementineRemote.Message;
+import de.qspool.clementineremote.backend.pb.ClementineRemote.MsgType;
+import de.qspool.clementineremote.backend.pb.ClementineRemote.RequestSetTrackPosition;
+import de.qspool.clementineremote.backend.pb.ClementineRemote.RequestSetVolume;
+import de.qspool.clementineremote.backend.pb.ClementineRemote.ResponseCurrentMetadata;
+import de.qspool.clementineremote.backend.pb.ClementineRemote.ResponseUpdateTrackPosition;
+import de.qspool.clementineremote.backend.pb.ClementineRemote.SongMetadata;
+
+/**
+ * The {@link ClementineRemoteHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * It uses Clementine's own, protocol buffers based communication sheme:
+ * https://github.com/clementine-player/Android-Remote/wiki/Developer-Documentation
+ *
+ * @author Stephan Richter - Initial contribution
+ */
+@NonNullByDefault
+public class ClementineRemoteHandler extends BaseThingHandler {
+
+    private int currentPos = 0;
+    private int volume = 50;
+
+    /**
+     * used to track the current state of the player
+     */
+    private enum State {
+        UNKNOWN,
+        PLAYING,
+        PAUSED,
+        SKIPPING,
+        STOPPED
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(ClementineRemoteHandler.class);
+
+    private @Nullable ClementineRemoteConfiguration config;
+    private @Nullable Socket socket = null;
+    private boolean enabled = true;
+    private @Nullable DataOutputStream out = null;
+    private @Nullable SongMetadata song = null;
+
+    private Message.Builder builder = Message.newBuilder();
+    private State state = State.UNKNOWN;
+
+    public ClementineRemoteHandler(Thing thing) {
+        super(thing);
+    }
+
+    /**
+     * Try to establish a connection to the Clementine instance.
+     * If the connection breaks: wait 15s, then try again
+     */
+    private void connect() {
+        enabled = true;
+        while (enabled) {
+            connectionStart();
+            if (enabled) {
+                logger.debug("Connection broken. Reconnectiong in 15 secs…");
+                sleep15secs();
+            }
+        }
+    }
+
+    /**
+     * Try to start a new connection:
+     * <ul>
+     * <li>open socket</li>
+     * <li>connect output stream</li>
+     * <li>send connect message</li>
+     * <li>wait for incoming messages</li>
+     * </ul>
+     * When the connection breaks: try to disconnect gracefully.
+     */
+    private void connectionStart() {
+        try {
+            if (socket == null || socket.isClosed()) {
```

Comment:
warning: Potential null pointer access. Better to use a local var

Socket socketLocal = socket;
if (socketLocal == null || socketLocal.isClosed()){

Reply:
Fixed in recent commit.